src/shogun/regression/KernelRidgeRegression.cpp

```diff
@@ -43,86 +42,60 @@ void CKernelRidgeRegression::init()
 	SG_ADD(&m_tau, "tau", "Regularization parameter", MS_AVAILABLE);
 }
 
-bool CKernelRidgeRegression::train_machine_pinv()
+bool CKernelRidgeRegression::solve_krr_system(SGVector<float64_t> alphas)
 {
-	// Get kernel matrix
-	SGMatrix<float64_t> kernel_matrix=kernel->get_kernel_matrix<float64_t>();
-	int32_t n = kernel_matrix.num_cols;
-	int32_t m = kernel_matrix.num_rows;
-	ASSERT(kernel_matrix.matrix && m>0 && n>0)
+	SGMatrix<float64_t> kernel_matrix(kernel->get_kernel_matrix());
+	int32_t n = kernel_matrix.num_rows;
+	SGVector<float64_t> y = ((CRegressionLabels*)m_labels)->get_labels();
 
-	for(int32_t i=0; i < n; i++)
-		kernel_matrix.matrix[i+i*n]+=m_tau;
-
-	/* re-set alphas of kernel machine */
-	m_alpha=((CRegressionLabels*) m_labels)->get_labels_copy();
-
-	/* tell kernel machine that all alphas are needed as'support vectors' */
-	m_svs=SGVector<index_t>(m_alpha.vlen);
-	m_svs.range_fill();
-
-	if (get_alphas().vlen!=n)
+	for(index_t i=0; i<n; i++)
 	{
-		SG_ERROR("Number of labels does not match number of kernel"
-				" columns (num_labels=%d cols=%d\n", m_alpha.vlen, n);
+		alphas[i] = y[i];
+		kernel_matrix(i,i) += m_tau;
 	}
 
-	clapack_dposv(CblasRowMajor,CblasUpper, n, 1, kernel_matrix.matrix, n,
-			m_alpha.vector, n);
+	Map<MatrixXd> eigen_kernel_matrix(kernel_matrix.matrix, n, n);
+	Map<VectorXd> eigen_alphas(alphas.vector, n);
 
+	LLT<MatrixXd> llt;
+	llt.compute(eigen_kernel_matrix);
+	if (llt.info() != Eigen::Success)
+	{
+		SG_WARNING("Features covariance matrix was not positive definite\n");
+		return false;
+	}
+	eigen_alphas = llt.solve(eigen_alphas);
 	return true;
 }
 
-bool CKernelRidgeRegression::train_machine_gs()
+bool CKernelRidgeRegression::train_machine(CFeatures *data)
 {
-	int32_t n = kernel->get_num_vec_rhs();
-	int32_t m = kernel->get_num_vec_lhs();
-	ASSERT(m>0 && n>0)
+	check_and_set_alphas(data);
```

Comment:
I'm calling this method here and not in the constructors because checks should only run when both labels and kernel are set which is not necessary in a constructor.